### PR TITLE
add tests for prototype builtin datasets

### PR DESCRIPTION
**Pull request by pmeier**
cc @pmeier @bjuncek

### CODE REVIEW

**Review by bjuncek — COMMENTED** [review]
Generally looks ok to me, but I'm slightly worried that this might add some overhead for users adding datasets no?

Also, is there a way to disable shufflers for testing? These tend to make things quite slow.

File: test/test_prototype_builtin_datasets.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    @builtin_datasets
+    def test_num_samples(self, dataset, mock_info):
+        num_samples = 0
+        for _ in dataset:
+            num_samples += 1
+
+        assert num_samples == mock_info["num_samples"]
+
+    @builtin_datasets
+    def test_decoding(self, dataset, mock_info):
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:71]: Good question. We should probably pass `decoder` here explicitly. I'll refactor.

File: test/test_prototype_builtin_datasets.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class TestCommon:
+    @builtin_datasets
+    def test_smoke(self, dataset, mock_info):
+        if not isinstance(dataset, IterDataPipe):
+            raise AssertionError(f"Loading the dataset should return an IterDataPipe, but got {type(dataset)} instead.")
+
+    @builtin_datasets
+    def test_sample(self, dataset, mock_info):
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:50]: I feel like it is reasonable to assume that if the first sample is valid, all the other will also be. Otherwise we would need to have some weird branching within the dataset that could cause this and this should be caught during code review.

Of course we can also  merge `test_sample` and `test_num_samples` to check the validity of every sample. But then the question arises, if we need to check all samples for every single test? For example, what about the decoding tests? Should we also check every sample there? If not, what is the difference to `test_sample`?

File: test/builtin_dataset_mocks.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+            return resources_and_mock_info
+
+        try:
+            fakedata_fn = self._mock_data_fns[name]
+        except KeyError:
+            raise pytest.UsageError(
+                f"No mock data available for dataset '{name}'. "
+                f"Did you add a new dataset, but forget to provide mock data for it? "
+                f"Did you register the mock data function with `@DatasetMocks.register_mock_data_fn`?"
+            )
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py:83]: > Perhaps I don't understand the decorators as well, but it looks like you put most of the mock data function here - won't this make this file huge and somewhat unreadable once we add more and more datasets?

Unfortunately, yes. 

> Should we perhaps add this mock data in the datasets themselves?

That is a fair question. In general I like the idea to have the mock data close the "actual" data. The problem I see, is that we would need to pull some pure test utilities such as `make_tar` or `make_image_folder` inside the `torchvision` package. I'm not opposed to that, but that would break the current paradigm that our tests are kind of "standalone".

PyTorch core also has a `testing` namespace with an `_internal` submodule where a lot of this functionality lives. We could also do that. In the light of #4721 we could also place some specialized comparison / generation functions there for downstream libraries to use. @NicolasHug

**Comment by pmeier** [comment]
> Generally looks ok to me, but I'm slightly worried that this might add some overhead for users adding datasets no?

Overhead as discussed in https://github.com/pytorch/vision/pull/4682#discussion_r741385339 or overhead in general, because the contributor also has to provide mock data along with the dataset?

If you mean the latter, than yes, this is "overhead", but it is well spent. If you have a look at the changes in this PR, I've found bugs in ImageNet, MNIST (and variants), as well as Caltech101 that were not caught be me or anyone else during code review. 

Plus, we currently also require a contributor to add tests including mock data for our legacy datasets API. In fact, IMO, it will be easier than it currently is, since the users only need to add a single function for the mock data, rather than provide a test case and override custom methods.

> Also, is there a way to disable shufflers for testing? These tend to make things quite slow.

I don't think the option is already available. IIRC, it will be available through the [`DataLoader`](https://github.com/pytorch/pytorch/blob/7f3326a6d2b51f5d38b0477f4a8444606ee2f203/torch/utils/data/dataloader_experimental.py#L58-L75). cc @ejguan @vitaly-fedyunin

**Comment by ejguan** [comment]
> I don't think the option is already available. IIRC, it will be available through the [`DataLoader`]

Correct. I think we will provide a switch in `DataLoader` to turn off all shufflers in the future. As a work around, you may want to only add `shuffler` in the pipeline when needed.

**Comment by pmeier** [comment]
@ejguan

> As a work around, you may want to only add shuffler in the pipeline when needed.

Given that one should never use a dataset without shuffling and shuffling has to happen inside our implementation. So unless we have a switch to turn it off through the data loader, we need to keep it in.

**Comment by ejguan** [comment]
> Given that one should never use a dataset without shuffling and shuffling has to happen inside our implementation. So unless we have a switch to turn it off through the data loader, we need to keep it in.

Understood. And DataLoader does have an option to optionally attach a `shuffler` at the end as a hack to turn on/off shuffling. The reason I didn't propose to you is global shuffling is required as we discussed. Shuffling at the end with an unlimited buffer is kind crazy especially considering all data are images.

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

Let me know if you want to get this merged now or wait to refactor following the comment from Bruno

**Comment by fmassa** [comment]
One thing to keep in mind, ideally, our datasets wouldn't be stochastic in the standard case (so `next(iter(ds))` always returns the same thing).
If users are always expected to use the dataset within a DataLoader, then fine. But this should be clear from the docs / etc

File: test/test_prototype_builtin_datasets.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjuncek) [reply test/test_prototype_builtin_datasets.py:50]: I feel like it's hard to do these without clogging up the CI. I'd say the first one is ok for 90% of the cases, but I think we should re-think these for the video datasets with clips to include somehow randomly drawn sample bc these can be a bit unstable.

[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:50]: If you are thinking about the video utilities you added in #4838, I don't think we need to change anything here. Since they are separate from the datasets, we can also test them separately.

File: test/builtin_dataset_mocks.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+            mock_resources.append(datasets.utils.LocalResource(path))
+
+        self._cache[(name, config)] = mock_resources, mock_info
+        return mock_resources, mock_info
+
+    def _decoder(self, dataset_type):
+        if dataset_type == datasets.utils.DatasetType.RAW:
+            return datasets.decoder.raw
+        else:
+            return lambda file: file.close()
```
Comment (pmeier) [inline test/builtin_dataset_mocks.py:106]: @bjuncek This is the decoder we are passing to each dataset during the tests unless we manually specify a different one. Thus, although the video datasets do not use a decoder by default through the normal API, they will use one during test.

File: test/test_prototype_builtin_datasets.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_prototype_builtin_datasets.py:71]: See https://github.com/pytorch/vision/pull/4682#discussion_r742629912.
